### PR TITLE
[storage] Add the ability to lookup reserved labels more efficiently.

### PR DESCRIPTION
**Pull request by esmet**
**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/query/storage/m3/storage.go [hunk 1]
```diff
@@ -127,29 +129,49 @@ func (s *m3storage) Name() string {
...
+	// The target should always be prefixed with reservedLabelPrefix
+	// but if it isn't at least that length, the answer is always empty.
+	if len(target) < len(reservedLabelPrefix) {
+		return nil
+	}
+
+	foundReservedLabels := false
+	for idx := 0; idx < len(labels); idx++ {
+		label := labels[idx]
+		if len(label.Name) < len(reservedLabelPrefix) || !bytes.Equal(label.Name[:2], reservedLabelPrefix) {
```
Comment (Aadeshp) [inline src/query/storage/m3/storage.go:144]: `s/:2/:len(reservedLabelPrefix)`? 😄

File: src/query/storage/m3/storage.go [hunk 2]
```diff
@@ -127,29 +129,52 @@ func (s *m3storage) Name() string {
...
+	if !bytes.HasPrefix(target, reservedLabelPrefix) {
+		return nil
+	}
+
+	foundReservedLabels := false
+	for idx := 0; idx < len(labels); idx++ {
+		label := labels[idx]
+		if !bytes.HasPrefix(label.Name, reservedLabelPrefix) {
+			if foundReservedLabels {
+				// We previously found reserved labels, and now that we've iterated
```
Comment (rallen090) [inline src/query/storage/m3/storage.go:147]: Ah we can do this because the labels are sorted. Makes sense

File: src/query/storage/m3/storage.go [hunk 3]
```diff
@@ -127,29 +129,52 @@ func (s *m3storage) Name() string {
...
-			} else if bytes.Equal(label.Name, promDefaultName) {
-				nameTag = label.Value
-			}
-		}
-
-		if rollup {
-			metadata.ByName(nameTag).Aggregated++
+		name := findReservedLabel(series.Labels, promDefaultName)
+		rollup := findReservedLabel(series.Labels, rollupTagName)
+		if bytes.Equal(rollup, rollupTagValue) {
```
Comment (rallen090) [inline src/query/storage/m3/storage.go:174]: Is this nil-safe?

**Review by rallen090 — APPROVED** [review]

File: src/query/storage/m3/storage.go (hunk 3 above)
Comment (rallen090) [reply src/query/storage/m3/storage.go:174]: Same w/ originally `nameTag`

Comment (esmet) [reply src/query/storage/m3/storage.go:174]: Yes bytes.Equal considers a nil argument to be equivalent to an empty slice. In practice it converts the argument to `string(...)`.

**Pull request merged**